README.md
# Raftify

[![CircleCI](https://circleci.com/gh/BlockscapeLab/raftify/tree/master.svg?style=shield)](https://circleci.com/gh/BlockscapeLab/raftify/tree/master)
[![codecov](https://codecov.io/gh/BlockscapeLab/raftify/branch/master/graph/badge.svg)](https://codecov.io/gh/BlockscapeLab/raftify)
[![Go Report Card](https://goreportcard.com/badge/github.com/blockscapelab/raftify)](https://goreportcard.com/report/github.com/blockscapelab/raftify)
[![License](https://img.shields.io/github/license/cosmos/cosmos-sdk.svg)](https://github.com/cosmos/cosmos-sdk/blob/master/LICENSE)

> :warning: This project has not yet had a security audit or stress test and is therefore not ready for use in production! Use at your own risk!

_Raftify_ is a Go implementation of the Raft leader election algorithm without the Raft log and enables the creation of a self-managing cluster of nodes by transforming an application into a Raft node. It is meant to be a **more cost-efficient** small-scale alternative to running a validator cluster with a separate full-fledged Raft consensus layer.

It is designed to be directly embedded into an application and provide a direct way of communicating between individual nodes, omitting the overhead caused by replicating a log.
Raftify was built with one particular use case in mind: **running a self-managing cluster of Cosmos validators**.

## Requirements

- Golang 1.14+

## Configuration Reference

The configuration is to be provided in a `raftify.json` file and must be located in the working directory specified in the second parameter of the `InitNode` method.

> :information_source: For Gaia, the working directory is `~/.gaiad/config/` by default.

| Key         | Value    | Description                                                                                                                                                                                                           |
|:------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `id`          | string   | **(Mandatory)** The node's identifier.</br>Must be **unique**.                                                                                                                                                         |
| `max_nodes`   | int      | **(Mandatory)** The self-imposed limit of nodes to be run in the cluster.</br>Must be greater than 0 and must _never_ be exceeded. |
| `expect`      | int      | **(Mandatory)** The number of nodes expected to be online in order to bootstrap the cluster and start the leader election. Once the expected number of nodes is online, all cluster members will be started simultaneously.</br>Must be 1 or higher and must _never_ exceed the self-imposed `max_nodes` limit.</br>:warning: Please use `expect = 1` for single-node setups only. If you plan on running more than one node, set the `expect` value to the final cluster size on **ALL** nodes. |
| `encrypt`     | string   | _(Optional)_ The hex representation of the secret key used to encrypt messages.</br>The value must be either 16, 24, or 32 bytes to select AES-128, AES-192, or AES-256.</br>[**Use this tool to generate a key.**](https://www.browserling.com/tools/random-bytes) |
| `performance` | int      | _(Optional)_ The modifier used to multiply the maximum and minimum timeout and ticker settings. Higher values increase leader stability and reduce bandwidth and CPU but also increase the time needed to recover from a leader failure.</br>Must be 1 or higher. Defaults to 1 which is also the maximum performance setting. |
| `log_level`   | string   | _(Optional)_ The minimum log level for console log messages.</br>Can be DEBUG, INFO, WARN, ERR. Defaults to `WARN`.                                                                                                    |
| `bind_addr`   | string   | _(Optional)_ The address to bind the node application to.</br>Defaults to `0.0.0.0`.                                                                                                                                                        |
| `bind_port`   | string   | _(Optional)_ The port to bind the node application to.</br>Defaults to `7946`.                                                                                                                                                              |
| `peer_list`   | []string | _(Optional)_ The list of IP addresses of all cluster members (optionally including the address of the local node). It is used to determine the quorum in a non-bootstrapped cluster.</br>For example, if your peerlist has `n = 3` nodes then `math.Floor((n/2)+1) = 2` nodes will need to be up and running to bootstrap the cluster.</br>Addresses must be provided in the `host:port` format.</br>Must not be empty if more than one node is expected. |

### Example Configuration

```json
{
    "id": "My-Unique-Name",
    "max_nodes": 3,
    "expect": 3,
    "encrypt": "8ba4770b00f703fcc9e7d94f857db0e76fd53178d3d55c3e600a9f0fda9a75ad",
    "performance": 1,
    "log_level": "WARN",
    "bind_addr": "192.168.0.25",
    "bind_port": 3000,
    "peer_list": [
        "192.168.0.25:3000",
        "192.168.0.26:3000",
        "192.168.0.27:3000"
    ]
}
```

## Getting Started

For a step-by-step guide on how to get started with your raftified Cosmos validator, check out [this tutorial](doc/getting-started.md).

## Testing

Use

```go
make unit-tests
```

to run unit tests, and

```go
make integration-tests
```

to run integration tests.


CHANGELOG.md
# Changelog

## Unreleased

---

\-

## v0.3.0

---

### General Changes

* Added custom handling for intended leave events in order to distinguish them better from crash-related ones

## v0.2.0

---

### Breaking Changes

* `InitNode` is now blocking. This means that on startup up a node, it will only unblock once the cluster is properly bootstrapped, that is if the expected number of nodes have formed a cluster

### Bugfixes

* Merge pull request #5: Fixed a bug that prevented nodes with `expect = 1` from becoming the cluster leader if there were other peers listed in the peerlist of the raftify.json file
* Merge pull request #9: Fixed a bug that prevented the prevote quorum from being adjusted to the new cluster size
* Merge pull request #10: Fixed a bug that caused a node to get stuck after a rejoin during operation

### General Changes

* Bump to memberlist `v0.2.2`
* Merge pull request #11: Added version info on startup

### Testing

* Merge pull request #7: Added more unit tests for more stable code coverage

## v0.1.0

---

* First release


config.go
package raftify

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"regexp"
	"strconv"
	"strings"

	"github.com/hashicorp/memberlist"

	"github.com/hashicorp/logutils"
)

// Timeout and ticker settings for maximum performance.
const (
	// Time interval measured in milliseconds in which candidates send out
	// vote requests and leaders send out heartbeats.
	TickerInterval = 200

	// Minimum time measured in milliseconds that a non-leader waits for a
	// current heartbeat to arrive before switching to a precandidate.
	MinTimeout = 800

	// Maximum time measured in milliseconds that a non-leader waits for a
	// current heartbeat to arrive before switching to a precandidate.
	MaxTimeout = 1200

	// Maximum number of cycles a leader can go without getting heartbeat
	// responses from a majority of cluster members. If there are not enough
	// heartbeat responses for more than cycles than this value, the leader
	// is forced to step down.
	MaxSubQuorumCycles = int(MinTimeout/TickerInterval) - 1

	// Maximum number of cycles a precandidate couldn't reach the quorum
	// in order to become a candidate. If this threshold is met or exceeded
	// it is safe to assume the cluster suffers a network partition and the
	// node in question is partitioned out into a smaller cluster that can
	// never reach the quorum. Upon reaching this threshold, a rejoin event
	// is triggered to make the node in question aware of the network partition.
	MaxMissedPrevoteCycles = 5
)

// Config contains the contents of the raftify.json file.
type Config struct {
	// Mandatory. The unique identifier of a node.
	ID string `json:"id"`

	// Mandatory. Self-imposed limit of nodes that can be run in one cluster.
	// This is needed to allocate enough memory for the buffered channel used
	// for event messages.
	MaxNodes int `json:"max_nodes"`

	// Mandatory. The 16-, 24- or 32-byte AES encryption key used to encrypt
	// the message exchange between cluster members.
	Encrypt string `json:"encrypt"`

	// The performance multiplier that determines how the timeouts and
	// intervals scale. This can be used to adjust the timeout settings
	// for higher latency environments.
	Performance int `json:"performance"`

	// The number of expected nodes to go online before starting the
	// Raft leader election and bootstrapping the cluster.
	Expect int `json:"expect"`

	// The log levels for raftify; can be DEBUG, INFO, WARN or ERR.
	LogLevel string `json:"log_level"`

	// The address to bind the node to.
	BindAddr string `json:"bind_addr"`

	// The port to bind the node to.
	BindPort int `json:"bind_port"`

	// The list of peers to contact in order to join an existing cluster
	// or form a new one.
	PeerList []string `json:"peer_list"`
}

// truncPeerList removes the local node from the peerlist.
func (c *Config) truncPeerList(address string) error {
	for i := range c.PeerList {
		if c.PeerList[i] == address {
			c.PeerList = append(c.PeerList[:i], c.PeerList[i+1:]...)
			return nil
		}
	}
	return errors.New("local node is not listed as a peer")
}

// validate checks for constraint violations in the raftify.json file.
func (c *Config) validate() error {
	// Variable used to aggregate all errors found during validation.
	var errs string

	// Set defaults.
	if c.Performance == 0 {
		c.Performance = 1
	}
	if c.LogLevel == "" {
		c.LogLevel = "WARN"
	}
	if c.BindAddr == "" {
		c.BindAddr = "0.0.0.0"
	}
	if c.BindPort == 0 {
		c.BindPort = 7946
	}

	// Check constraints.
	if c.ID == "" {
		errs += "\tid must not be empty\n"
	}
	if c.MaxNodes <= 0 {
		errs += "\tmax_nodes must be greater than 0\n"
	}
	if secretKey, err := hexToByte(c.Encrypt); err != nil {
		if err := memberlist.ValidateKey(secretKey); err != nil {
			errs += fmt.Sprintf("\tencrypt must be of length 16, 24 or 32 bytes: got %v bytes\n", len(secretKey))
		}
	}
	if c.Performance < 0 {
		errs += "\tperformance must be greater than 0\n"
	}
	if c.Expect < 1 || c.Expect > c.MaxNodes {
		errs += fmt.Sprintf("\texpect must be between 1 and %v\n", c.MaxNodes)
	}
	if c.Expect > 1 && len(c.PeerList) == 0 {
		errs += "\tpeerlist must not be empty if more than one node is expected for bootstrap\n"
	}
	if match, _ := regexp.MatchString(`DEBUG|INFO|WARN|ERR`, c.LogLevel); !match {
		errs += "\tlog_level must be DEBUG, INFO, WARN or ERR\n"
	}
	if ip := net.ParseIP(c.BindAddr); ip == nil {
		errs += "\tbind_addr is not a valid IPv4\n"
	}
	if c.BindPort < 0 || c.BindPort > 65535 {
		errs += fmt.Sprintf("\tbind_port %v must be in range 0-65535\n", c.BindPort)
	}
	if len(c.PeerList) > c.MaxNodes {
		errs += fmt.Sprintf("\tpeer_list must not contain more than %v peers, including the local node: got %v peers\n", c.MaxNodes, len(c.PeerList))
	}
	for _, peer := range c.PeerList {
		host, port, err := net.SplitHostPort(peer)
		if err != nil {
			errs += fmt.Sprintf("\tpeer address %v is not a valid host:port address\n", peer)
		}
		if ip := net.ParseIP(host); ip == nil {
			errs += fmt.Sprintf("\tbind_addr %v is not a valid IPv4\n", host)
		}
		if p, _ := strconv.Atoi(port); p < 0 || p > 65535 {
			errs += fmt.Sprintf("\tbind_port %v must be in range 0-65535\n", port)
		}
	}

	if errs != "" {
		return fmt.Errorf("found errors in raftify.json:\n%v", strings.TrimSuffix(errs, "\n"))
	}
	return nil
}

// loadConfig loads the contents of the raftify.json file into memory.
func (n *Node) loadConfig(stateJSONExists bool) error {
	configJSON, err := os.Open(n.workingDir + "/raftify.json")
	if err != nil {
		return err
	}
	defer configJSON.Close()

	configBytes, err := ioutil.ReadAll(configJSON)
	if err != nil {
		return err
	}
	if err = json.Unmarshal(configBytes, &n.config); err != nil {
		return err
	}

	// If the state.json file exists, overwrite the peerlist from the raftify.json
	// with the memberlist persisted in the state.json file.
	if stateJSONExists {
		n.logger.Println("[DEBUG] raftify: Overwriting peerlist with peers from state.json...")

		list, err := n.loadState()
		if err != nil {
			return err
		}

		n.config.PeerList = []string{}
		localNode := fmt.Sprintf("%v:%v", n.config.BindAddr, n.config.BindPort)

		for _, node := range list {
			if node.Address() == localNode {
				continue
			}
			n.config.PeerList = append(n.config.PeerList, node.Address())
		}
	}

	// Remove local node from peerlist such that the join event throws an error if none of
	// the other peers could be reached. It needs to be removed because it will always reach
	// itself which is obvious on one hand and not necessary on the other.
	// Also, the truncation needs to be done before the validation.
	n.config.truncPeerList(fmt.Sprintf("%v:%v", n.config.BindAddr, n.config.BindPort))

	if err = n.config.validate(); err != nil {
		return err
	}

	n.logger.SetOutput(&logutils.LevelFilter{
		Levels:   []logutils.LogLevel{"DEBUG", "INFO", "WARN", "ERR"},
		MinLevel: logutils.LogLevel(n.config.LogLevel),
		Writer:   os.Stderr,
	})
	return nil
}
